### PR TITLE
Add --with-lg-tcache-limit to increase the max supported tcaches

### PR DESCRIPTION
**Pull request by veselink1**

### CODE REVIEW

File: include/jemalloc/internal/jemalloc_internal_types.h.in [hunk 1]
```diff
@@ -33,17 +33,18 @@ typedef enum malloc_init_e malloc_init_t;
  *
  * a: arena
  * t: tcache
+ * .: configuration-dependent
```
Comment (daverigby) [inline include/jemalloc/internal/jemalloc_internal_types.h.in]: Suggest calling out the name of the config param which controls this.

File: configure.ac [hunk 2]
```diff
@@ -1671,6 +1671,37 @@ if test "x${je_cv_gcc_builtin_popcountl}" = "xyes" ; then
...
+AC_ARG_WITH([lg_tcache_limit],
+  [AS_HELP_STRING([--with-lg-tcache-limit=<lg-cache-limit>],
+  [Base 2 log of maximum number of tcaches (between 12 and 15). Trade off with number of arenas, which will be 2 ^ (24 - lg-cache-limit).])])
+
+if test "x$with_lg_tcache_limit" = "x"; then
+  # Default is 4096 tcaches, 4096 arenas
+  with_lg_tcache_limit=12
+fi
+
+if ! test "$with_lg_tcache_limit" -ge 12 -a "$with_lg_tcache_limit" -le 15; then
```
Comment (daverigby) [inline configure.ac]: This feels over-specified to me. If we are giving the user (of jemalloc) the choice of how many tcaches / arenas they want, why stop them having more than 15 (or less than 12?)

Suggest allowing it to be configured up to the limit of what jemalloc fundamentally supports - i.e. maybe at least 1 bit for arena and at least one for tcache?

File: configure.ac [hunk 3]
```diff
@@ -1671,6 +1671,37 @@ if test "x${je_cv_gcc_builtin_popcountl}" = "xyes" ; then
...
+  AC_MSG_ERROR([lg-cache-limit should be between 12 and 15.])
+fi
+
+dnl Internal macros
+tcache_and_arena_repr_bits=24
+tcache_repr_bits="$with_lg_tcache_limit"
+arena_repr_bits="$(($tcache_and_arena_repr_bits-$tcache_repr_bits))"
+tcache_shift=8
```
Comment (daverigby) [inline configure.ac:1698]: I'd minimise the amount of maths we do in autoconf - it's not a particulary nice place to do that. 

Just expose the number of bits for tcache, and do all the other calculations in C.

**Review by daverigby — CHANGES_REQUESTED** [review]

**Review by daverigby — COMMENTED** [review]
Suggest you ask the jemalloc devs what they prefer as to where the macros are defined - for example they might prefer to keep the existing autoconf'd files and just move the 1-2 macro definitions which need configuring into say  https://github.com/couchbasedeps/jemalloc/blob/82caac43b38eda300a54b1e28327a60ed7235c1f/include/jemalloc/jemalloc_defs.h.in

File: include/jemalloc/internal/jemalloc_internal_types.h.in (hunk 1 above)
Comment (veselink1) [reply include/jemalloc/internal/jemalloc_internal_types.h.in]: Done

File: configure.ac (hunk 2 above)
Comment (veselink1) [reply configure.ac]: Having less than some number of arenas (17 with my unit test changes) makes some of the unit tests fail.

File: configure.ac (hunk 3 above)
Comment (veselink1) [reply configure.ac:1698]: Agree about it not being nice, however, we need the arenas_all and arenas_destroyed to be literals. 

See the comment on:
https://github.com/couchbasedeps/jemalloc/blob/5.3.0-couchbase/include/jemalloc/jemalloc_macros.h.in#L44

File: configure.ac [hunk 4]
```diff
@@ -1671,6 +1671,41 @@ if test "x${je_cv_gcc_builtin_popcountl}" = "xyes" ; then
   AC_DEFINE([JEMALLOC_INTERNAL_POPCOUNTLL], [__builtin_popcountll], [ ])
 fi
 
+dnl Configure maximum number of tcaches/arenas
+dnl The upper limit of 17 is not a technical limitation, rather, it is what our
+dnl unit tests pass at. Some tests allocate areans and never destroy them, which
+dnl means we run out of areans during tests. But we can't "support" a limit we
+dnl cannot test...
```
Comment (daverigby) [inline configure.ac:1680]: Thanks. 

Suggest also documenting the lower bound of 4.

File: include/jemalloc/internal/jemalloc_internal_types.h.in [hunk 5]
```diff
@@ -33,17 +33,18 @@ typedef enum malloc_init_e malloc_init_t;
  *
  * a: arena
  * t: tcache
+ * .: dependent on --with-lg-tcache-limit
  * 0: unused
  * z: zero
  * n: alignment
  *
- * aaaaaaaa aaaatttt tttttttt 0znnnnnn
+ * aaaaaaaa a...tttt tttttttt 0znnnnnn
```
Comment (daverigby) [inline include/jemalloc/internal/jemalloc_internal_types.h.in]: I think this needs updating now, given you're allowing a larger range of values in the config param.

File: test/unit/background_thread_enable.c [hunk 6]
```diff
@@ -12,9 +12,13 @@ max_test_narenas(void) {
...
 	 */
 	unsigned ret = 10 * ncpus;
-	/* Limit the max to avoid VM exhaustion on 32-bit . */
-	if (ret > 512) {
-		ret = 512;
+	// Limit the max to avoid running out of arenas. There are quite a few
+	// tests where arenas.create is called but arena.<i>.destroy is not, so the
+	// arena remains. Once those are fixed, we should be able to change this to
+	// something like MIN(512, ARENAS_LIMIT).
+	// (Cap at 512 to avoid VM exhaustion on 32-bit.)
```
Comment (daverigby) [inline test/unit/background_thread_enable.c]: This line (19) doesn't match the code now.

**Review by daverigby — CHANGES_REQUESTED** [review]

File: test/unit/background_thread_enable.c (hunk 6 above)
Comment (veselink1) [reply test/unit/background_thread_enable.c]: Done, thanks.

File: include/jemalloc/internal/jemalloc_internal_types.h.in (hunk 5 above)
Comment (veselink1) [reply include/jemalloc/internal/jemalloc_internal_types.h.in]: Done

File: configure.ac (hunk 4 above)
Comment (veselink1) [reply configure.ac:1680]: Done

File: include/jemalloc/internal/jemalloc_internal_types.h.in [hunk 7]
```diff
@@ -33,17 +33,18 @@ typedef enum malloc_init_e malloc_init_t;
  *
  * a: arena
  * t: tcache
+ * .: dependent on --with-lg-tcache-limit
```
Comment (daverigby) [inline include/jemalloc/internal/jemalloc_internal_types.h.in]: Nit: maybe make explicit in the comment this is either (a)rena or (t)cache - at the moment it's not clear what `.` is actually going to be used for.

**Review by daverigby — CHANGES_REQUESTED** [review]

File: include/jemalloc/internal/jemalloc_internal_types.h.in (hunk 7 above)
Comment (veselink1) [reply include/jemalloc/internal/jemalloc_internal_types.h.in]: Done, thanks.

**Review by daverigby — APPROVED** [review]

**Pull request merged**